File: backend/app/schemas/team.py
import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field

from app.models.enums import GameType, MemberRole


class TeamCreate(BaseModel):
    name: str = Field(..., min_length=2, max_length=100)
    tag: str = Field(..., min_length=2, max_length=10, pattern=r"^[A-Za-z0-9]+$")
    game: GameType
    description: Optional[str] = Field(None, max_length=1000)
    country: Optional[str] = Field(None, max_length=100)
    logo_url: Optional[str] = Field(None, max_length=2048)
    banner_url: Optional[str] = Field(None, max_length=2048)
    twitter_handle: Optional[str] = Field(None, max_length=50)


class TeamUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=2, max_length=100)
    tag: Optional[str] = Field(None, min_length=2, max_length=10, pattern=r"^[A-Za-z0-9]+$")
    description: Optional[str] = Field(None, max_length=1000)
    country: Optional[str] = Field(None, max_length=100)
    logo_url: Optional[str] = Field(None, max_length=2048)
    banner_url: Optional[str] = Field(None, max_length=2048)
    twitter_handle: Optional[str] = Field(None, max_length=50)


class TeamSummarySchema(BaseModel):
    id: uuid.UUID
    name: str
    tag: str
    game: str
    logo_url: Optional[str]
    owner_id: uuid.UUID
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class TeamDetailSchema(TeamSummarySchema):
    description: Optional[str]
    country: Optional[str]
    banner_url: Optional[str]
    twitter_handle: Optional[str]
    updated_at: datetime


class TeamMemberSchema(BaseModel):
    id: uuid.UUID
    player_id: uuid.UUID
    user_id: Optional[uuid.UUID]
    in_game_name: Optional[str]
    username: Optional[str]
    avatar_url: Optional[str]
    role: str
    jersey_number: Optional[int]
    joined_at: datetime
    is_active: bool


class AddMemberRequest(BaseModel):
    username: str = Field(..., description="招待するユーザーのユーザー名")
    role: MemberRole = MemberRole.PLAYER
    jersey_number: Optional[int] = Field(None, ge=1, le=99)


class UpdateMemberRoleRequest(BaseModel):
    role: MemberRole

File: frontend/src/types/team.ts
import type { GameType } from "./tournament";

export type TeamRole = "player" | "coach" | "analyst" | "manager" | "substitute";

export interface Team {
  id: string;
  owner_id: string;
  name: string;
  tag: string;
  logo_url: string | null;
  banner_url: string | null;
  game: GameType;
  region: string | null;
  country: string | null;
  founded_at: string | null;
  description: string | null;
  social_links: {
    twitter?: string;
    twitch?: string;
    discord?: string;
    website?: string;
  } | null;
  rating: number;
  created_at: string;
  updated_at: string;
}

export interface TeamMember {
  id: string;
  player_id: string;
  user_id: string | null;
  username: string | null;
  display_name: string | null;
  avatar_url: string | null;
  role: TeamRole;
  in_game_name: string | null;
  jersey_number: number | null;
  agent_specialty: string[] | null;
  joined_at: string;
  is_active: boolean;
  rating: number | null;
}

export interface TeamStats {
  total_matches: number;
  wins: number;
  losses: number;
  win_rate: number;
  game_wins: number;
  game_losses: number;
  game_win_rate: number;
  rating: number;
  peak_rating: number;
  tournaments_played: number;
  tournaments_won: number;
  current_streak: number;
  best_streak: number;
}

export interface WinRatePoint {
  month: string;
  win_rate: number;
  rating: number;
  matches: number;
}

export interface MonthlyStatsPoint {
  month: string;
  wins: number;
  losses: number;
  tournaments: number;
}

export interface TeamAnalytics {
  win_rate_history: WinRatePoint[];
  monthly_stats: MonthlyStatsPoint[];
}

export interface TeamMatchSummary {
  id: string;
  tournament_id: string;
  tournament_name: string;
  opponent: {
    id: string;
    name: string;
    tag: string;
    logo_url: string | null;
  } | null;
  result: "win" | "loss" | null;
  team_score: number;
  opponent_score: number;
  status: string;
  played_at: string | null;
  format: string;
}
